pkg/didcomm/protocol/introduce/service.go

```diff
@@ -424,52 +477,67 @@ func (s *Service) handle(msg *metaData, dest *service.Destination) error {
 	logger.Infof("next valid state to transition -> %s ", next.Name())
 
 	for !isNoOp(next) {
-		s.sendMsgEvents(&service.StateMsg{
-			ProtocolName: Introduce,
-			Type:         service.PreState,
-			Msg:          msg.Msg.Clone(),
-			StateID:      next.Name(),
-		})
-		logger.Infof("sent pre event for state %s", next.Name())
-
-		var (
-			followup state
-			err      error
-		)
-
-		if dest != nil {
-			followup, err = next.ExecuteOutbound(s.ctx, msg, dest)
-		} else {
-			followup, err = next.ExecuteInbound(s.ctx, msg)
-		}
-
+		next, err = s.execute(next, msg, dest)
 		if err != nil {
-			return fmt.Errorf("execute state %s %w", next.Name(), err)
+			return fmt.Errorf("execute: %w", err)
 		}
+	}
 
-		logger.Infof("finish execute next state: %s", next.Name())
+	return nil
+}
 
-		if err = s.save(msg.ThreadID, record{
-			StateName: next.Name(),
-			WaitCount: msg.WaitCount,
-		}); err != nil {
-			return fmt.Errorf("failed to persist state %s %w", next.Name(), err)
-		}
+func (s *Service) execute(next state, msg *metaData, dest *service.Destination) (state, error) {
+	s.sendMsgEvents(&service.StateMsg{
+		ProtocolName: Introduce,
+		Type:         service.PreState,
+		Msg:          msg.Msg.Clone(),
+		StateID:      next.Name(),
+	})
+	logger.Infof("sent pre event for state %s", next.Name())
 
-		logger.Infof("persisted the connection using %s and updated the state to %s", msg.ThreadID, next.Name())
+	var (
+		followup state
+		err      error
+	)
 
-		s.sendMsgEvents(&service.StateMsg{
-			ProtocolName: Introduce,
-			Type:         service.PostState,
-			Msg:          msg.Msg.Clone(),
-			StateID:      next.Name(),
-		})
-		logger.Infof("sent post event for state %s", next.Name())
+	if err = injectInvitation(msg); err != nil {
+		return nil, fmt.Errorf("inject invitation: %w", err)
+	}
 
-		next = followup
+	if dest != nil {
+		followup, err = next.ExecuteOutbound(s.ctx, msg, dest)
+	} else {
+		followup, err = next.ExecuteInbound(s.ctx, msg)
 	}
 
-	return nil
+	if err != nil {
+		return nil, fmt.Errorf("execute state %s %w", next.Name(), err)
+	}
+
+	logger.Infof("finish execute next state: %s", next.Name())
+
+	r := record{
+		StateName:       next.Name(),
+		WaitCount:       msg.WaitCount,
+		IntroduceeIndex: msg.IntroduceeIndex,
+		Invitation:      msg.Invitation,
+	}
+
+	if err = s.save(msg.ThreadID, r); err != nil {
```

Comment:
Currently, introducees cannot respond at the same time. It works consistently. Send proposal -> wait for a response ->  send another proposal ->  wait for a response etc. The problem here can be the same as we had in did-exchange. When we send a message and we received a response before saving the current state. It can be fixed by moving save function (before state execution). I am gonna fix it latter. Just need to think about how it can be fixed in a better way.

Reply:
It is also a problem that we cannot send both proposals at once. This should be optimized in a followup. Please create the issue :).

Reply:
https://github.com/hyperledger/aries-framework-go/issues/902